rust/feature-flags/src/cohort_operations.rs

```diff
@@ -0,0 +1,369 @@
+    pub fn extract_dependencies(&self) -> Result<HashSet<CohortId>, FlagError> {
+        let cohort_property: CohortProperty = serde_json::from_value(self.filters.clone())
+            .map_err(|e| {
+                tracing::error!("Failed to parse filters for cohort {}: {}", self.id, e);
+                FlagError::CohortFiltersParsingError
+            })?;
+
+        let mut dependencies = HashSet::new();
+        Self::traverse_filters(&cohort_property.properties, &mut dependencies)?;
+        Ok(dependencies)
+    }
+
+    /// Recursively traverses the filter tree to find cohort dependencies
+    ///
+    /// Example filter tree structure:
+    /// ```json
+    /// {
+    ///   "properties": {
+    ///     "type": "OR",
+    ///     "values": [
+    ///       {
+    ///         "type": "OR",
+    ///         "values": [
+    ///           {
+    ///             "key": "id",
+    ///             "value": 123,
+    ///             "type": "cohort",
+    ///             "operator": "exact"
+    ///           },
+    ///           {
+    ///             "key": "email",
+    ///             "value": "@posthog.com",
+    ///             "type": "person",
+    ///             "operator": "icontains"
+    ///           }
+    ///         ]
+    ///       }
+    ///     ]
+    ///   }
+    /// }
+    /// ```
+    fn traverse_filters(
+        inner: &InnerCohortProperty,
+        dependencies: &mut HashSet<CohortId>,
+    ) -> Result<(), FlagError> {
+        for cohort_values in &inner.values {
+            for filter in &cohort_values.values {
+                if filter.is_cohort() {
+                    // Assuming the value is a single integer CohortId
+                    if let Some(cohort_id) = filter.value.as_i64() {
+                        dependencies.insert(cohort_id as CohortId);
+                    } else {
+                        return Err(FlagError::CohortFiltersParsingError);
+                    }
+                }
+                // NB: we don't support nested cohort properties, so we don't need to traverse further
+            }
+        }
+        Ok(())
+    }
+}
+
+impl InnerCohortProperty {
+    /// Flattens the nested cohort property structure into a list of property filters.
+    ///
+    /// The cohort property structure in Postgres looks like:
+    /// ```json
+    /// {
+    ///   "type": "OR",
+    ///   "values": [
+    ///     {
+    ///       "type": "OR",
+    ///       "values": [
+    ///         {
+    ///           "key": "email",
+    ///           "value": "@posthog.com",
+    ///           "type": "person",
+    ///           "operator": "icontains"
+    ///         },
+    ///         {
+    ///           "key": "age",
+    ///           "value": 25,
+    ///           "type": "person",
+    ///           "operator": "gt"
+    ///         }
+    ///       ]
+    ///     }
+    ///   ]
+    /// }
+    /// ```
+    pub fn to_property_filters(&self) -> Vec<PropertyFilter> {
```

Comment:
Just checked - you can consume the `self` here and nothing breaks, there's nowhere you want to both get a cloned copy of the inner filters and keep the wrapper around. This function becomes:
```rust
    pub fn to_property_filters(self) -> Vec<PropertyFilter> {
        self.values
            .into_iter()
            .flat_map(|value| value.values)
            .collect()
    }
```

And I'd call it something like `to_inner()`. A semantic note but in rust, `to_` prefixed functions almost always consume the `self`, whereas `as` or `get` ones take a `&str` (the `to_` implying no allocation).